# 异常机制

## 异常结构

![img](https://imgconvert.csdnimg.cn/aHR0cHM6Ly9pbWctbXkuY3Nkbi5uZXQvdXBsb2Fkcy8yMDEyMTEvMjcvMTM1NDAyMDQxN181MTc2LmpwZw?x-oss-process=image/format,png)

**Error（错误）:**是**程序无法处理的错误**，表示运行应用程序中较严重问题。大多数错误与代码编写者执行的操作无关，而表示代码运行时 JVM（Java 虚拟机）出现的问题。例如OutOfMemoryError。这些异常发生时，Java虚拟机（JVM）一般会选择线程终止。

**Exception（异常）:**是程序本身可以处理的异常。

**运行时异常**：RuntimeException类及其子类，如NullPointerException(空指针异常)、1/0异常等

**非运行时异常：**是RuntimeException以外的异常。从程序语法角度讲是必须进行处理的异常，如果不处理，程序就不能编译通过。如IOException、SQLException等以及用户自定义的Exception异常，一般情况下不自定义检查异常。

**checked Exceptions** :非运行时异常

**unchecked Exceptions** 包括ERROR和RuntimeException

## 处理异常机制

 **抛出异常**：自己解决不了，抛出给上层解决

 **捕获异常**：捕获并解决异常

**try 块：**用于捕获异常。其后可接零个或多个catch块，如果没有catch块，则必须跟一个finally块。
**catch 块：**用于处理try捕获到的异常。
**finally 块：**无论是否捕获或处理异常，finally块里的语句都会被执行。当在try块或catch块中遇到return语句时，finally语句块将在方法返回之前被执行。当finally中有return语句时，执行finally中的return语句，就不管其它try中的return了，即finally外的return不可达了。

在以下4种特殊情况下，finally块不会被执行：
1）在finally语句块中发生了异常。
2）在前面的代码中用了System.exit()退出程序。
3）程序所在的线程死亡。
4）关闭CPU。

**Error能被catch吗？**

可以，因为是Throwable子类，但是ERROR一般是严重的问题，不建议catch捕获

## 自定义异常类

**1）创建自定义异常类：**一般会选择继承Exception和RuntimeException，如果不要求调用者一定要处理抛出的异常，就继承RuntimeException。
**2）抛出自定义异常：**在方法中通过throw关键字抛出异常对象。
**3）捕获自定义异常：**如果在当前抛出异常的方法中处理异常，可以使用try-catch语句捕获并处理；否则在方法的声明处通过throws关键字指明要抛出给方法调用者的异常，继续进行下一步操作。
4）在出现异常方法的调用者中捕获并处理异常

## 原理

在jvm中对于执行的方法都会有一个执行方法栈区，记录相应的数据包括类名、方法名、文件名、调用的行数，每次执行的时候jvm会生成一张栈表用于记录执行的每一行代码的执行，当代码发生异常的时候，会回溯代码的执行栈以及抛出相对应的代码位置，代码嵌套的越多回溯的时间越长，性能影响较大

## 问题

**StackOverFlow原因**

栈溢出，一般是死循环，死递归，造成栈数量过大。

局部变量体积过大，比如定义一个大数组，导致栈帧太大。

**OutOfMemory原因**

`Java heap space` ，堆溢出，创建对象太多，可分配的堆内存已满

`PermGen space` 永久代溢出，永久代一般存放类信息、常量池、类名、访问修饰符等。